File: codeforces/1900~1999/1943/b.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

vector<int> manacher(const string& s) {
  string t = "^#";
  for (char ch : s) {
    t.push_back(ch);
    t.push_back('#');
  }
  t.push_back('$');

  int n = t.length();
  vector<int> p(n);
  int l = 1, r = 1;
  for (int i = 1; i < n-1; ++i) {
    p[i] = max(0, min(r-i, p[l+r-i]));
    while (t[i + p[i]] == t[i - p[i]]) {
      p[i]++;
    }
    if (i + p[i] > r) {
      r = i + p[i];
      l = i - p[i];
    }
  }
  return vector<int>(p.begin()+1, p.end()-1);
}

void solve() {
  int n = 0, q = 0;
  cin >> n >> q;
  string s;
  cin >> s;

  vector<vector<int>> chars(26, vector<int>(n+1));
  for (int i = 1; i <= n; ++i) {
    int c = s[i-1] - 'a';
    chars[c][i]++;
  }
  for (int c = 0; c < 26; ++c) {
    for (int i = 1; i <= n; ++i) {
      chars[c][i] += chars[c][i-1];
    }
  }

  auto is_all_same = [&](int l, int r) -> bool {
    int len = r - l + 1;
    for (int c = 0; c < 26; ++c) {
      int count = chars[c][r+1] - chars[c][l];
      if (count != 0 && count != len) {
        return false;
      }
    }
    return true;
  };

  vector<int> p = manacher(s);

  auto is_palindrome = [&](int l, int r) -> bool {
    int len = r - l + 1;
    int palindrome_len = p[l * 2 + len] - 1;
    // cout << "p=" << palindrome_len << ", " << len << endl;
    return palindrome_len >= len;
  };

  vector<int> next_diff(n, n);
  vector<int> st;
  for (int i = 0; i < n; i += 2) {
    while (!st.empty() && s[st.back()] != s[i]) {
      next_diff[st.back()] = i;
      st.pop_back();
    }
    st.push_back(i);
  }
  st.clear();
  for (int i = 1; i < n; i += 2) {
    while (!st.empty() && s[st.back()] != s[i]) {
      next_diff[st.back()] = i;
      st.pop_back();
    }
    st.push_back(i);
  }

  auto is_alternating = [&](int l, int r) -> bool {
    int i1 = next_diff[l];
    int i2 = next_diff[l+1];
    // cout << "next diff: " << i1 << " " << i2 << endl;
    return i1 > r && i2 > r;
  };

  for (int i = 0; i < q; ++i) {
    int l = 0, r = 0;
    cin >> l >> r;
    i64 ans = 0;
    l--;
    r--;
    int len = r - l;
    if (len >= 3 && !is_alternating(l, r)) {
      i64 l = len % 2 == 1 ? len : len-1;
      // cout << "add from 3 ~ " << l << endl;
      ans += (l + 3) * ((l - 3) / 2 + 1) / 2;
    }
    if (len >= 2 && !is_all_same(l, r)) {
      i64 l = len % 2 == 0 ? len : len-1;
      // cout << "add from 2 ~ " << l << endl;
      ans += (l + 2) * ((l - 2) / 2 + 1) / 2;
    }

    len = r - l + 1;
    if (!is_palindrome(l, r)) {
      // cout << "add " << len << endl;
      ans += len;
    }
    cout << ans << endl;
  }
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1900~1999/1943/b.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn read_str() -> String {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn manacher(s: &Vec<char>) -> Vec<i32> {
    let mut t = Vec::new();
    t.push('^');
    t.push('#');
    for &ch in s.iter() {
        t.push(ch);
        t.push('#');
    }
    t.push('$');

    let n = t.len();
    let mut p = vec![0; n];
    let mut l: i32 = 1;
    let mut r: i32 = 1;
    for i in 1..(n-1) {
        p[i] = (r - i as i32).min(p[(l + r) as usize - i]);
        p[i] = p[i].max(0);
        while t[i + p[i] as usize] == t[i -  p[i] as usize] {
            p[i] += 1;
        }
        if i as i32 + p[i] > r {
            l = i as i32 - p[i];
            r = i as i32 + p[i];
        }
    }
    p.into_iter().skip(1).take(n-1).collect()
}

fn is_palindrome(l: usize, r: usize, p: &Vec<i32>) -> bool {
    let len = r - l + 1;
    let palindrome_len = p[l * 2 + len] as usize - 1;
    palindrome_len >= len
}

fn compute_next_diff(s: &Vec<char>) -> Vec<usize> {
    let n = s.len();
    let mut next_diff = vec![n; n];
    let mut st = Vec::new();
    for i in (0..n).step_by(2) {
        while !st.is_empty() && s[st[st.len()-1]] != s[i] {
            next_diff[st[st.len()-1]] = i;
            st.pop();
        }
        st.push(i);
    }
    st.clear();
    for i in (1..n).step_by(2) {
        while !st.is_empty() && s[st[st.len()-1]] != s[i] {
            next_diff[st[st.len()-1]] = i;
            st.pop();
        }
        st.push(i);
    }
    next_diff
}

fn is_alternating(l: usize, r: usize, next_diff: &Vec<usize>) -> bool {
    let i1 = next_diff[l];
    let i2 = next_diff[l+1];
    i1 > r && i2 > r
}

fn compute_char_counts(s: &Vec<char>) -> Vec<Vec<i32>> {
    let n = s.len();
    let mut chars = vec![vec![0; n+1]; 26];
    for i in 1..=n {
        let c = (s[i-1] as u8 - b'a') as usize;
        chars[c][i] += 1;
    }
    for c in 0..26 {
        for i in 1..=n {
            chars[c][i] += chars[c][i-1];
        }
    }
    chars
}

fn is_all_same(l: usize, r: usize, chars: &Vec<Vec<i32>>) -> bool {
    let len = (r - l + 1) as i32;
    for c in 0..26 {
        let count = chars[c][r+1] - chars[c][l];
        if count != 0 && count != len {
            return false;
        }
    }
    true
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let inputs = read_vec::<usize>();
    let q = inputs[1];
    let s: Vec<char> = read_str().chars().collect();

    let p = manacher(&s);
    let next_diff = compute_next_diff(&s);
    let chars = compute_char_counts(&s);

    for _i in 0..q {
        let query = read_vec::<usize>();
        let l = query[0] - 1;
        let r = query[1] - 1;
        let mut ans: i64 = 0;

        let mut len = (r - l) as i64;
        if len >= 3 && !is_alternating(l, r, &next_diff) {
            let l = if len % 2 == 1 { len } else { len - 1 };
            ans += (l + 3) * ((l - 3) / 2 + 1) / 2;
        }

        if len >= 2 && !is_all_same(l, r, &chars) {
            let l = if len % 2 == 0 { len } else { len - 1 };
            ans += (l + 2) * ((l - 2) / 2 + 1) / 2;
        }

        len = (r - l + 1) as i64;
        if !is_palindrome(l, r, &p) {
            ans += len;
        }
        writeln!(writer, "{}", ans).ok();
    }
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
